hivemind/utils/asyncio.py

```diff
@@ -114,23 +114,30 @@ async def amap_in_executor(
     queue = asyncio.Queue(max_prefetch)
 
     async def _put_items():
-        async for args in azip(*iterables):
-            await queue.put(loop.run_in_executor(executor, func, *args))
-        await queue.put(None)
+        try:
+            async for args in azip(*iterables):
+                await queue.put(loop.run_in_executor(executor, func, *args))
+            await queue.put(None)
+        except BaseException as e:
+            await queue.put(e)  # note: there is no chance that iterables
+            raise
 
     task = asyncio.create_task(_put_items())
     try:
-        future = await queue.get()
-        while future is not None:
-            yield await future
-            future = await queue.get()
+        future_or_exception = await queue.get()
+        while future_or_exception is not None:
+            if isinstance(future_or_exception, BaseException):
+                raise future_or_exception
+            yield await future_or_exception
+            future_or_exception = await queue.get()
         await task
     finally:
-        if not task.done():
-            task.cancel()
+        task.cancel()
+        if task.done() and not task.cancelled():
+            task.exception()
```

Comment:
I specifically mean "if task did not send result or throw exception but we died anyway, silence the "task ... was never retrieved""

Reply:
Let's remove L136-137 because `.cancel()` already suppresses this message.

Indeed, the message is printed only when `task.__log_traceback == True` ([source](https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Lib/asyncio/futures.py#L94)) and `task.cancel()` sets it to `False` ([source](https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Lib/asyncio/futures.py#L151)), exactly as `task.exception()` does ([source](https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Lib/asyncio/futures.py#L217)).